runtime/src/accounts_cache.rs

```diff
@@ -0,0 +1,236 @@
+use dashmap::DashMap;
+use solana_sdk::{account::Account, clock::Slot, hash::Hash, pubkey::Pubkey};
+use std::{
+    collections::HashSet,
+    ops::Deref,
+    sync::{
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc, RwLock,
+    },
+};
+
+pub type SlotCache = Arc<SlotCacheInner>;
+
+#[derive(Default, Debug)]
+pub struct SlotCacheInner {
+    cache: DashMap<Pubkey, CachedAccount>,
+    same_account_writes: AtomicU64,
+    same_account_writes_size: AtomicU64,
+    unique_account_writes_size: AtomicU64,
+    is_frozen: AtomicBool,
+}
+
+impl SlotCacheInner {
+    pub fn report_slot_store_metrics(&self) {
+        datapoint_info!(
+            "slot_repeated_writes",
+            (
+                "same_account_writes",
+                self.same_account_writes.load(Ordering::Relaxed),
+                i64
+            ),
+            (
+                "same_account_writes_size",
+                self.same_account_writes_size.load(Ordering::Relaxed),
+                i64
+            ),
+            (
+                "unique_account_writes_size",
+                self.unique_account_writes_size.load(Ordering::Relaxed),
+                i64
+            )
+        );
+    }
+
+    pub fn insert(&self, pubkey: &Pubkey, account: Account, hash: Hash) {
```

Comment:
I wonder how it's complicated and/or beneficial to compute a hash for given account only once per slot at the end for delta_hash.

Reply:
in other words, I think we can delay computing it as much as possible.

Reply:
hmmm yeah I was worried it would delay freezing, I think ideally you have the hash computed asynchronously over the course of the block.

Reply:
oh, I meant intra-slot (= intra-bank) optimization. So, N multiple updates to the same account in a single slot/bank can avoid computing hash of N different versions. I think we need only latest version hashs for delta hash of the bank/slot for voting?

Anyway, this could be a follow-up pr or just pend. Already, this pr is big. :)